File: 0392_is_subsequence/clojure/src/solution_followup.clj
(ns solution-followup
  (:import [java.util Collections]))

(defn build-index-map
  "Preprocess t into a map of char -> sorted vector of indices."
  [t]
  (reduce-kv
   (fn [m i ch] (update m ch (fnil conj []) i))
   {}
   (vec t)))

(defn- bisect-left
  "Find the first index in sorted coll where value >= target."
  [coll target]
  (let [idx (Collections/binarySearch coll (int target))]
    (if (neg? idx)
      (- (inc idx))  ; insertion point
      idx)))

(defn is-subsequence-followup
  "Check if s is a subsequence using precomputed index-map."
  [index-map s]
  (loop [pos 0
         chars (seq s)]
    (if (nil? chars)
      true
      (let [ch (first chars)
            indices (get index-map ch)]
        (if (nil? indices)
          false
          (let [idx (bisect-left indices pos)]
            (if (>= idx (count indices))
              false
              (recur (inc (nth indices idx)) (next chars)))))))))

File: 0392_is_subsequence/kotlin/solution_followup.kt
import java.util.Collections

class SubsequenceChecker(t: String) {
    private val indexMap = mutableMapOf<Char, MutableList<Int>>()

    init {
        for ((index, char) in t.withIndex()) {
            indexMap.getOrPut(char) { mutableListOf() }.add(index)
        }
    }

    fun isSubsequence(s: String): Boolean {
        var position = 0

        for (char in s) {
            val indices = indexMap[char] ?: return false
            val rawIndex = Collections.binarySearch(indices, position)
            val insertionPoint = if (rawIndex >= 0) rawIndex else -rawIndex - 1
            if (insertionPoint == indices.size) return false
            position = indices[insertionPoint] + 1
        }

        return true
    }
}
